### ISSUE TITLE
[Python] Issues after "Update to 3.9.7" commit because of a buggy patch

### ISSUE THREAD

**Issue by Jehan** [msys2/MINGW-packages#9677]
In GIMP, we have a weekly Windows build (we also have others at each commits, but these don't run the GObject Introspection rules because they are cross-builds and gobject-introspection tools are not working when cross-compiling). Anyway the last native pipeline happened yesterday and it failed with some GObject Introspection script.

Here is a recent log with `make V=1` (verbose log): https://gitlab.gnome.org/GNOME/gimp/-/jobs/1540677
The build fails with:

```sh
dllwrap: WARNING: dllwrap is deprecated, use gcc -shared or ld -shared instead
Traceback (most recent call last):
  File "C:\msys64\mingw64\bin\g-ir-scanner-script.py", line 100, in <module>
    sys.exit(scanner_main(sys.argv))
  File "C:/msys64/mingw64/lib/gobject-introspection/giscanner/scannermain.py", line 591, in scanner_main
    ss, filenames = create_source_scanner(options, args)
  File "C:/msys64/mingw64/lib/gobject-introspection/giscanner/scannermain.py", line 463, in create_source_scanner
    ss.parse_files(filenames)
  File "C:/msys64/mingw64/lib/gobject-introspection/giscanner/sourcescanner.py", line 261, in parse_files
    self._parse(headers)
  File "C:/msys64/mingw64/lib/gobject-introspection/giscanner/sourcescanner.py", line 293, in _parse
    cc = CCompiler()
  File "C:/msys64/mingw64/lib/gobject-introspection/giscanner/ccompiler.py", line 140, in __init__
    self.compiler = distutils.ccompiler.new_compiler(compiler=compiler_name)
  File "C:/msys64/mingw64/lib/python3.9/distutils/ccompiler.py", line 1034, in new_compiler
    return klass(None, dry_run, force)
  File "C:/msys64/mingw64/lib/python3.9/distutils/cygwinccompiler.py", line 331, in __init__
    if is_cygwincc(self.cc):
  File "C:/msys64/mingw64/lib/python3.9/distutils/cygwinccompiler.py", line 451, in is_cygwincc
    out_string = check_output([cc, '-dumpmachine'])
  File "C:/msys64/mingw64/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:/msys64/mingw64/lib/python3.9/subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:/msys64/mingw64/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:/msys64/mingw64/lib/python3.9/subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
make[3]: *** [C:/msys64/mingw64/share/gobject-introspection-1.0/Makefile.introspection:158: Gimp-3.0.gir] Error 1
make[3]: Leaving directory '/c/_r/_builds/k3_3muaB/0/GNOME/gimp/_build-w64/libgimp'
make[2]: *** [Makefile:1658: all] Error 2
make[2]: Leaving directory '/c/_r/_builds/k3_3muaB/0/GNOME/gimp/_build-w64/libgimp'
make[1]: *** [Makefile:875: all-recursive] Error 1
make[1]: Leaving directory '/c/_r/_builds/k3_3muaB/0/GNOME/gimp/_build-w64'
make: *** [Makefile:776: all] Error 2
```

It's probably not the `dllwrap` warning which is the problem because I think we had it before. The real error is not very clear though (`FileNotFoundError` but it doesn't say which file. Probably the answer would be in the code of these package's scripts).

I see that the `gobject-introspection` package was very recently updated (10 days ago) so the issue might come from the new version. But I couldn't find any recent report. Any help appreciated.

**Comment by jeremyd2019** [comment]
The error is coming out of the `CreateProcess` call, and the call stack shows `out_string = check_output([cc, '-dumpmachine'])` so I imagine `cc` is not found.  I would be concerned about a relocation issue with python there

**Comment by jeremyd2019** [comment]
No, this should all be very straightforward.  `cc` should be `os.environ.get('CC', 'gcc')`

**Comment by Jehan** [comment]
So a bug in the package? We have not changed a thing in our build relative to GIR, and the stuff is all generic autotools detection. 🙂

**Comment by Jehan** [comment]
Ok so thanks to @jeremyd2019's diagnosis, I found the issue. We were indeed setting `export CC="ccache gcc"`. While this is not the best way to set `ccache`, it is not wrong either and use to work until now.

Since `distutils/cygwinccompiler.py`, the file where the issue happens, belongs to the "python3" module, I thought at first the issue came from a recent change in Python 3.9.7, as I see MSYS2 recently updated. This is indeed the case, but it is not an upstream bug! Indeed upstream file in the `3.9` branch [still has](https://github.com/python/cpython/blob/2cf76cf4ccd177b8d6d2bf21b5462258ae87522d/Lib/distutils/cygwinccompiler.py#L402):

```Python
    out_string = check_output(['gcc', '-dumpmachine'])
```

In other words, the bug would not happen with upstream Python. It happens only in MSYS2 because of the patch `mingw-w64-python/0097-distutils-get-compilers-from-env-vars.patch` added in recent commit 8198ab6550b7bcb98838333f2facb8c1aab9d022:

```diff
-def is_cygwingcc():
-    '''Try to determine if the gcc that would be used is from cygwin.'''
-    out_string = check_output(['gcc', '-dumpmachine'])
+def is_cygwincc(cc):
+    '''Try to determine if the compiler that would be used is from cygwin.'''
+    out_string = check_output([cc, '-dumpmachine'])
```

@lazka Where does this patch comes from? CPython repository says about the [associated commit](https://github.com/python/cpython/commit/b1f1cbeefcc5ed956ee484cd96008d5d37eb7db2) that "*This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.*" so it is not even a backport.

In any case, this patch creates problem. Maybe it also fixes some actual issue (apparently using clang instead of gcc) but not the correct way.

**Comment by lazka** [comment]
This is fixed since "3.9.7-3" via https://github.com/msys2/MINGW-packages/pull/10048
